spec/type_references_and_ownership.md

```diff
@@ -0,0 +1,331 @@
+      Each segment of the name corresponds to an owning resource, so creating a Batch Pool `foo` in Batch account `account` would have `name` = `account/foo`.
+
+#### What ASO does
+Dependent resources in ASO have properties which map to the name/path to their owner. For example 
+[MySQLFirewallRuleSpec](https://github.com/Azure/azure-service-operator/blob/92240406aff3863f3a267d8a1dc1e28aa3e841ae/api/v1alpha1/mysqlfirewallrule_types.go#L15) looks like this:
+
+```go
+type MySQLFirewallRuleSpec struct {
+	ResourceGroup  string `json:"resourceGroup"`
+	Server         string `json:"server"`
+	StartIPAddress string `json:"startIpAddress"`
+	EndIPAddress   string `json:"endIpAddress"`
+}
+```
+
+The `ResourceGroup` and `Server` are references to the owners of this type.
+
+#### What k8s-infra does
+k8s-infra uses the same `KnownTypeReference` type mentioned above for ownership references too.
+There are two patterns for ownership in k8s-infra today.
+
+One pattern is used for ResourceGroup, where 
+[top level resources have a link to the resource group they are in](https://github.com/Azure/k8s-infra/blob/14105b1cb3f6967cd086c9f8f75fb16bb85d6318/apis/microsoft.compute/v1/virtualmachinescaleset_types.go#L323).
+
+```go
+type VirtualMachineScaleSetSpec struct {
+    // +kubebuilder:validation:Required
+    ResourceGroupRef *azcorev1.KnownTypeReference `json:"resourceGroupRef" group:"microsoft.resources.infra.azure.com" kind:"ResourceGroup"`
+    ...
+}
+```
+
+The other pattern is where the owning resource has links to the dependent resources it is expecting to have:
+
+```go
+type RouteTableSpecProperties struct {
+    DisableBGPRoutePropagation bool                          `json:"disableBgpRoutePropagation,omitempty"`
+    RouteRefs                  []azcorev1.KnownTypeReference `json:"routeRefs,omitempty" group:"microsoft.network.infra.azure.com" kind:"Route" owned:"true"`
+}
+```
+
+If the dependent resources aren't there, the status of the owning resource reflects an error explaining that.
+
+## Proposal
+
+### How to represent references
+We construct a new type which is our reference holder. I am calling this approach `AugmentedKnownTypeReference` (real type name TBD), because
+it's similar to the current approach that `k8s-infra` is taking but with a few changes.
+
+```go
+type AugmentedKnownTypeReference struct {
+    Kubernetes KubernetesReference
+    Azure      AzureReference
+}
+
+type KubernetesReference struct {
+    // The "Group" of the owning resources GVK. There may be an annotation on the property 
+    // that specifies Group, in which case this is optional (if supplied it must match the group
+    // on the annotation). For resources which do not have a group annotation, this is required.
+    // TODO: It's possibly going to be confusing to customers when they have to specify this and when they don't...
```

Comment:
I was worried about this being confusing that is why there is [`KnownTypeReference`](https://github.com/Azure/k8s-infra/blob/317e01fb2bbcb1e1b74d346cc1f470c8307f9514/apis/core/v1/types.go#L39-L45) and [`TypedReference`](https://github.com/Azure/k8s-infra/blob/317e01fb2bbcb1e1b74d346cc1f470c8307f9514/apis/core/v1/types.go#L27-L37) in k8s-infra. The latter includes the group and kind, where as the former expects its included in the associated metadata of the struct tag. Turns out we usually know the type based on the location of the references.

Reply:
Ah yeah, two different types might be a good idea.

> Turns out we usually know the type based on the location of the references.
Yeah based on my hacking so far, for _all_ ownership (parent -> child) relationships we can autogenerate the group and kind as metadata in the field tag.

For just "references" (i.e. VMSS -> Subnet) it's a bit trickier both to identify the properties in question as well as know for sure what the types are supposed to be.

Reply:
This is in the list of open questions now (should ownership and references use the same type or different types).

I tend you agree with you @devigned that it would be clearer if there were two types, so I'll tentatively plan on updating the specification to include that. Will leave this comment unresolved though so others can see it and weigh in.

Reply:
Have updated the spec to default to this approach (two types)